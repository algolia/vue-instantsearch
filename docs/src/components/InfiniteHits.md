---
title: InfiniteHits
mainTitle: Components
layout: main.pug
category: Components
withHeadings: true
navWeight: 6
editable: true
githubSource: docs/src/components/InfiniteHits.md
---

Displays an infinite list of hits along with a load more button.

<a class="btn btn-static-theme" href="stories/?selectedKind=InfiniteHits">ðŸ•¹ try out live</a>

## Usage

```html
<ais-infinite-hits></ais-infinite-hits>
```

## Props

None.

## CSS classes

Here's a list of CSS classes exposed by this widget. To better understand the underlying
DOM structure, have a look at the generated DOM in your browser.

Class name | Description
---|---
`ais-InfiniteHits` | The root div of the widget
`ais-InfiniteHits-list` | The list of hits
`ais-InfiniteHits-item` | The hit list item
`ais-InfiniteHits-loadMore` | The button used to display more results
`ais-InfiniteHits-loadMore--disabled` | The disabled button used to display more results
