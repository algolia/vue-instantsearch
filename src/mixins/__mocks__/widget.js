let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
