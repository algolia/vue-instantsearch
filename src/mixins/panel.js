import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
