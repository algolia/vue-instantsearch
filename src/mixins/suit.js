import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
