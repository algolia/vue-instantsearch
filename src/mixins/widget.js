const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
