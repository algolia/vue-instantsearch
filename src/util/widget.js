export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
