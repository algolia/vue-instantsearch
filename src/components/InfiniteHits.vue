<template>
  <div
    v-if="state"
    :class="suit()"
  >
    <slot
      :items="items"
      :is-last-page="state.isLastPage"
      :refine="refine"
    >
      <ol :class="suit('list')">
        <li
          v-for="(item, index) in items"
          :class="suit('item')"
          :key="item.objectID"
        >
          <slot
            name="item"
            :item="item"
            :index="index"
          >
            objectID: {{ item.objectID }}, index: {{ index }}
          </slot>
        </li>
      </ol>

      <button
        :class="[suit('loadMore'), state.isLastPage && suit('loadMore', 'disabled')]"
        :disabled="state.isLastPage"
        @click="refine"
      >
        Show more results
      </button>
    </slot>
  </div>
</template>

<script>
import { createWidgetMixin } from '../mixins/widget';
import { connectInfiniteHits } from 'instantsearch.js/es/connectors';

export default {
  mixins: [createWidgetMixin({ connector: connectInfiniteHits })],
  props: {
    escapeHTML: {
      type: Boolean,
      default: true,
    },
    transformItems: {
      type: Function,
      default(items) {
        return items;
      },
    },
  },
  data() {
    return {
      widgetName: 'InfiniteHits',
    };
  },
  computed: {
    widgetParams() {
      return {
        escapeHits: this.escapeHTML,
        transformItems: this.transformItems,
      };
    },
    items() {
      // Fixes InstantSearch.js connectors API: every list
      // of things must be called `items`
      return this.state.hits;
    },
  },
  methods: {
    refine() {
      // Fixes InstantSearch.js connectors API: every function
      // that trigger a search must be called `refine`
      this.state.showMore();
    },
  },
};
</script>
