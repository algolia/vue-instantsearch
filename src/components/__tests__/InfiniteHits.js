import { mount } from '@vue/test-utils';
import { __setState } from '../../mixins/widget';
import InfiniteHits from '../InfiniteHits.vue';

jest.mock('../../mixins/widget');

const defaultState = {
  widgetParams: {
    showPrevious: false,
    escapeHTML: true,
    transformItems: items => items,
  },
  hits: [
    {
      objectID: '00001',
    },
    {
      objectID: '00002',
    },
    {
      objectID: '00003',
    },
    {
      objectID: '00004',
    },
    {
      objectID: '00005',
    },
  ],
  results: {
    page: 0,
  },
  isFirstPage: false,
  isLastPage: false,
  showMore: () => {},
  showPrevious: () => {},
};

it('accepts a escapeHTML prop', () => {
  __setState({
    ...defaultState,
  });

  const wrapper = mount(InfiniteHits, {
    propsData: {
      escapeHTML: true,
    },
  });

  expect(wrapper.vm.widgetParams.escapeHTML).toBe(true);
});

it('accepts a transformItems prop', () => {
  __setState({
    ...defaultState,
  });

  const transformItems = () => {};

  const wrapper = mount(InfiniteHits, {
    propsData: {
      transformItems,
    },
  });

  expect(wrapper.vm.widgetParams.transformItems).toBe(transformItems);
});

it('accepts a showPrevious prop', () => {
  __setState({
    ...defaultState,
  });

  const wrapper = mount(InfiniteHits, {
    propsData: {
      showPrevious: true,
    },
  });

  expect(wrapper.vm.widgetParams.showPrevious).toBe(true);
});

it('renders correctly', () => {
  __setState({
    ...defaultState,
  });

  const wrapper = mount(InfiniteHits);

  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly with a custom rendering', () => {
  __setState({
    ...defaultState,
  });

  const wrapper = mount(InfiniteHits, {
    scopedSlots: {
      default: `
        <div slot-scope="{ items }">
          <div v-for="item in items">
            {{item.objectID}}
          </div>
        </div>
      `,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly with a custom item rendering', () => {
  __setState({
    ...defaultState,
  });

  const wrapper = mount(InfiniteHits, {
    scopedSlots: {
      item: `
        <div slot-scope="{ item }">
          {{item.objectID}}
        </div>
      `,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly on the first page', () => {
  __setState({
    ...defaultState,
    isFirstPage: true,
  });

  const wrapper = mount(InfiniteHits, {
    propsData: {
      showPrevious: true,
    },
  });

  const previousButton = wrapper.find('.ais-InfiniteHits-loadPrevious');

  expect(previousButton.exists()).toEqual(true);
  expect(
    previousButton.classes('ais-InfiniteHits-loadPrevious--disabled')
  ).toEqual(true);
  expect(previousButton.attributes('disabled')).toEqual('disabled');
  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly on the last page', () => {
  __setState({
    ...defaultState,
    isLastPage: true,
  });

  const wrapper = mount(InfiniteHits);

  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly when not on the first page', () => {
  __setState({
    ...defaultState,
    isFirstPage: false,
    isLastPage: false,
  });

  const wrapper = mount(InfiniteHits, {
    propsData: {
      showPrevious: true,
    },
  });

  const previousButton = wrapper.find('.ais-InfiniteHits-loadPrevious');

  expect(previousButton.exists()).toEqual(true);
  expect(
    previousButton.classes('ais-InfiniteHits-loadPrevious--disabled')
  ).toEqual(false);
  expect(previousButton.attributes('disabled')).toEqual(undefined);
  expect(wrapper.html()).toMatchSnapshot();
});

it('expect to call showPrevious on click', () => {
  const showPrevious = jest.fn();

  __setState({
    ...defaultState,
    showPrevious,
  });

  const wrapper = mount(InfiniteHits, {
    propsData: {
      showPrevious: true,
    },
  });

  expect(showPrevious).toHaveBeenCalledTimes(0);

  wrapper.find('.ais-InfiniteHits-loadPrevious').trigger('click');

  expect(showPrevious).toHaveBeenCalledTimes(1);
});

it('expect to call showMore on click', () => {
  const showMore = jest.fn();

  __setState({
    ...defaultState,
    showMore,
  });

  const wrapper = mount(InfiniteHits);

  expect(showMore).not.toHaveBeenCalled();

  wrapper.find('.ais-InfiniteHits-loadMore').trigger('click');

  expect(showMore).toHaveBeenCalled();
});

it('exposes insights prop to the default slot', () => {
  const insights = jest.fn();

  __setState({
    ...defaultState,
    insights,
  });

  const wrapper = mount(InfiniteHits, {
    scopedSlots: {
      default: `
        <ul slot-scope="{ items, insights }">
          <li v-for="(item, itemIndex) in items" >
            <button :id="'add-to-cart-' + item.objectID" @click="insights('clickedObjectIDsAfterSearch', {eventName: 'Add to cart', objectIDs: [item.objectID]})">
              Add to cart
            </button>
          </li>
        </ul>
      `,
    },
  });
  wrapper.find('#add-to-cart-00002').trigger('click');
  expect(insights).toHaveBeenCalledWith('clickedObjectIDsAfterSearch', {
    eventName: 'Add to cart',
    objectIDs: ['00002'],
  });
});

it('exposes insights prop to the item slot', () => {
  const insights = jest.fn();

  __setState({
    ...defaultState,
    insights,
  });

  const wrapper = mount(InfiniteHits, {
    scopedSlots: {
      item: `
          <div slot-scope="{ item, insights }">
            <button :id="'add-to-cart-' + item.objectID" @click="insights('clickedObjectIDsAfterSearch', {eventName: 'Add to cart', objectIDs: [item.objectID]})">
              Add to cart
            </button>
          </div>
      `,
    },
  });
  wrapper.find('#add-to-cart-00002').trigger('click');
  expect(insights).toHaveBeenCalledWith('clickedObjectIDsAfterSearch', {
    eventName: 'Add to cart',
    objectIDs: ['00002'],
  });
});
