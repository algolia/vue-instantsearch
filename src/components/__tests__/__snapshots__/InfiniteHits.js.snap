// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `

<div class="ais-InfiniteHits">
  <ol class="ais-InfiniteHits-list">
    <li class="ais-InfiniteHits-item">
      objectID: 00001, index: 0
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00002, index: 1
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00003, index: 2
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00004, index: 3
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00005, index: 4
    </li>
  </ol>
  <button class="ais-InfiniteHits-loadMore">
    Show more results
  </button>
</div>

`;

exports[`renders correctly on the first page 1`] = `

<div class="ais-InfiniteHits">
  <button disabled="disabled"
          class="ais-InfiniteHits-loadPrevious ais-InfiniteHits-loadPrevious--disabled"
  >
    Show previous results
  </button>
  <ol class="ais-InfiniteHits-list">
    <li class="ais-InfiniteHits-item">
      objectID: 00001, index: 0
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00002, index: 1
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00003, index: 2
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00004, index: 3
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00005, index: 4
    </li>
  </ol>
  <button class="ais-InfiniteHits-loadMore">
    Show more results
  </button>
</div>

`;

exports[`renders correctly on the last page 1`] = `

<div class="ais-InfiniteHits">
  <ol class="ais-InfiniteHits-list">
    <li class="ais-InfiniteHits-item">
      objectID: 00001, index: 0
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00002, index: 1
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00003, index: 2
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00004, index: 3
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00005, index: 4
    </li>
  </ol>
  <button disabled="disabled"
          class="ais-InfiniteHits-loadMore ais-InfiniteHits-loadMore--disabled"
  >
    Show more results
  </button>
</div>

`;

exports[`renders correctly when not on the first page 1`] = `

<div class="ais-InfiniteHits">
  <button class="ais-InfiniteHits-loadPrevious">
    Show previous results
  </button>
  <ol class="ais-InfiniteHits-list">
    <li class="ais-InfiniteHits-item">
      objectID: 00001, index: 0
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00002, index: 1
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00003, index: 2
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00004, index: 3
    </li>
    <li class="ais-InfiniteHits-item">
      objectID: 00005, index: 4
    </li>
  </ol>
  <button class="ais-InfiniteHits-loadMore">
    Show more results
  </button>
</div>

`;

exports[`renders correctly with a custom item rendering 1`] = `

<div class="ais-InfiniteHits">
  <ol class="ais-InfiniteHits-list">
    <li class="ais-InfiniteHits-item">
      <div>
        00001
      </div>
    </li>
    <li class="ais-InfiniteHits-item">
      <div>
        00002
      </div>
    </li>
    <li class="ais-InfiniteHits-item">
      <div>
        00003
      </div>
    </li>
    <li class="ais-InfiniteHits-item">
      <div>
        00004
      </div>
    </li>
    <li class="ais-InfiniteHits-item">
      <div>
        00005
      </div>
    </li>
  </ol>
  <button class="ais-InfiniteHits-loadMore">
    Show more results
  </button>
</div>

`;

exports[`renders correctly with a custom rendering 1`] = `

<div class="ais-InfiniteHits">
  <div>
    <div>
      00001
    </div>
    <div>
      00002
    </div>
    <div>
      00003
    </div>
    <div>
      00004
    </div>
    <div>
      00005
    </div>
  </div>
</div>

`;
